kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/CoroutineLoggingInterceptor.kt

```diff
@@ -0,0 +1,170 @@
+package io.kotest.engine.test.interceptors
+
+import io.kotest.common.ExperimentalKotest
+import io.kotest.core.config.LogLevel
+import io.kotest.core.config.configuration
+import io.kotest.core.extensions.Extension
+import io.kotest.core.test.TestCase
+import io.kotest.core.test.TestContext
+import io.kotest.core.test.TestResult
+import io.kotest.engine.test.withCoroutineContext
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.coroutines.AbstractCoroutineContextElement
+import kotlin.coroutines.CoroutineContext
+
+data class LogEntry(val level: LogLevel, val message: Any)
+
+/**
+ * An extension that is invoked when a test completes, logging any values that were logged using
+ * by way of [trace], [debug], [info], [warn], and [error] during that test.
+ *
+ * Users can use testId on [TestCase.description] to cross-reference the [TestResult] with the provided logs.
+ */
+@ExperimentalKotest
+interface LogExtension : Extension {
+   suspend fun handleLogs(testCase: TestCase, logs: List<LogEntry>)
+}
+
+/**
+ * [SerialLogExtension] wraps a [LogExtension] with a mutex, so we can guarantee
+ * that calls to [LogExtension.handleLogs] are invoked sequentially.
+ */
+@ExperimentalKotest
+internal class SerialLogExtension constructor(private val logExtension: LogExtension) {
+   private val mutex = Mutex()
+
+   suspend fun handleLogs(testCase: TestCase, logs: List<LogEntry>) = mutex.withLock {
+      runCatching {
+         logExtension.handleLogs(testCase, logs)
+      }
+   }
+}
+
+/**
+ * Returns the [TestLogger] that is embedded with this [TestContext].
+ * Does not error when there isn't a logger, there isn't one when LogLevel is set to Off.
+ */
+@ExperimentalKotest
+internal val TestContext.logs: MutableList<LogEntry>?
+   get() = coroutineContext[TestContextLoggingCoroutineContextElement]?.logs
+
+@ExperimentalKotest
+internal class TestContextLoggingCoroutineContextElement(val logs: MutableList<LogEntry>) : AbstractCoroutineContextElement(Key) {
+   companion object Key : CoroutineContext.Key<TestContextLoggingCoroutineContextElement>
+}
+
+@OptIn(ExperimentalKotest::class)
+internal object CoroutineLoggingInterceptor : TestExecutionInterceptor {
+   override suspend fun intercept(
+      test: suspend (TestCase, TestContext) -> TestResult
+   ): suspend (TestCase, TestContext) -> TestResult = { testCase, context ->
+      val extensions = configuration.extensions().filterIsInstance<LogExtension>().map { SerialLogExtension(it) }
+
+      when {
+         configuration.logLevel.isDisabled() || extensions.isEmpty() -> test(testCase, context)
+         else -> {
+            val logs = mutableListOf<LogEntry>()
+            val contextWithLogging = context.withCoroutineContext(TestContextLoggingCoroutineContextElement(logs))
+            try {
+               test(testCase, contextWithLogging)
```

Comment:
You might want to wrap test(testCase, contextWithLogging) with a withContext call, because otherwise, downstream interceptors can change it without realizing.
I will fix this up.